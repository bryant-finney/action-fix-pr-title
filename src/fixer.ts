/**
 * @class Fixer
 * If possible, correct improperly formatted issue keys in the given string.
 *
 * @param prefixes A list of prefixes
 * @param raw The string to check and fix
 */
class Fixer {
  prefixes: string[]
  raw: string
  re: RegExp

  constructor(prefixes: string[], raw: string) {
    this.prefixes = prefixes
    this.raw = raw
    this.re = new RegExp(
      `(?<![a-z])(${prefixes.join('|')})[^A-Z0-9](\\d+)(?![A-Z])`,
      'gi'
    )
  }

  /**
   * Return a single regex match from the PR title.
   *
   * When multiple matches are found, each invocation returns the subsequent match.
   *
   * @returns An array of issue keys or null if no matches are found.
   */
  check(): string[] | null {
    return this.re.exec(this.raw)
  }

  /**
   * Correct improperly formatted issue keys.
   *
   * @returns The corrected string.
   */
  fix(): string {
    return this.raw.replace(this.re, (_, prefix: string, number: string) =>
      prefix.toUpperCase().concat('-', number)
    )
  }
}

/**
 * Fix the issue keys in the PR title.
 *
 * @param prefixes An array of valid issue prefix strings
 * @param title The title of the PR to fix
 *
 * @returns The fixed PR title
 * @throws {Error} If no issue keys are found
 */
async function fix(prefixes: string[], title: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const fixer = new Fixer(prefixes, title)
    if (!fixer.check())
      return reject(new Error(`No issue keys found: ${fixer.raw}`))

    return resolve(fixer.fix())
  })
}

export { Fixer, fix }
