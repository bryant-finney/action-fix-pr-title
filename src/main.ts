import * as core from '@actions/core'
import { fix } from './fixer'

/**
 * The main function for the action.
 * @returns {Promise<void>} Resolves when the action is complete.
 */
export async function run(): Promise<void> {
  try {
    const title: string = core.getInput('title')
    const prefixes: string[] = core.getInput('prefixes').split(',')

    const fixed: string = await fix(prefixes, title)

    core.setOutput('title', fixed)
    core.setOutput('fixed', fixed !== title)
  } catch (error) {
    if (error instanceof Error) core.setFailed(error.message)
  }
}
