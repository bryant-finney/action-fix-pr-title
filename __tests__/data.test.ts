/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { Fixer } from '../src/fixer'
import { data } from './data.json'

const prefixes = ['JIRA']
const repo = 'test-0'

describe('data', () => {
  it(`includes a repo named ${repo}`, () => {
    expect(data.organization.repository.name).toEqual(repo)
  })
})

// TODO: add more data
describe('fix() when run on data', () => {
  const titles = data.organization.repository.pullRequests.nodes
    .map(pr => pr.title)
    .map(title => new Fixer(prefixes, title).fix())

  it('corrects the PR titles', () => {
    expect(titles).toMatchSnapshot()
  })
})
