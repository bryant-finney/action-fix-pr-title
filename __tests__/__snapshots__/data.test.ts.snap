// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fix() when run on data corrects the PR titles 1`] = `
[
  "feat/JIRA-5",
  "Dependabot PR to upgrade dependencies",
]
`;
